isisd/isis_sr.c

```diff
@@ -430,108 +622,203 @@ static mpls_label_t isis_sr_prefix_out_label(const struct sr_prefix *srp,
+				"%s: couldn't install Prefix-SID %pFX: loopback interface not found",
+				__func__, &srp->prefix);
+			return;
+		}
+
+		znh = &zl.nexthops[zl.nexthop_num++];
+		znh->type = NEXTHOP_TYPE_IFINDEX;
+		znh->ifindex = ifp->ifindex;
+		znh->label_num = 1;
+		/* Always POP incoming label even if PHP is requested */
+		znh->labels[0] = MPLS_LABEL_IMPLICIT_NULL;
+		break;
+	case ISIS_SR_PREFIX_REMOTE:
+		/* Update route in the RIB too. */
+		SET_FLAG(zl.message, ZAPI_LABELS_FTN);
+		zl.route.prefix = srp->prefix;
+		zl.route.type = ZEBRA_ROUTE_ISIS;
+		zl.route.instance = 0;
+
+		/* Add all potential route for ECMP support */
+		for (ALL_LIST_ELEMENTS_RO(srp->u.remote.rinfo->nexthops, node,
+					  nexthop)) {
+			if (nexthop->sr.label == MPLS_INVALID_LABEL)
+				continue;
+
+			if (zl.nexthop_num >= MULTIPATH_NUM)
+				break;
+
+			znh = &zl.nexthops[zl.nexthop_num++];
+			znh->type = (srp->prefix.family == AF_INET)
+					    ? NEXTHOP_TYPE_IPV4_IFINDEX
+					    : NEXTHOP_TYPE_IPV6_IFINDEX;
+			znh->gate = nexthop->ip;
+			znh->ifindex = nexthop->ifindex;
+			znh->label_num = 1;
+			znh->labels[0] = nexthop->sr.label;
+		}
+		break;
+	}
+
+	/* Send message to zebra. */
+	(void)zebra_send_mpls_labels(zclient, ZEBRA_MPLS_LABELS_REPLACE, &zl);
+}
+
+/**
+ * Uninstall Prefix-SID from the forwarding plane through Zebra.
+ *
+ * @param srp	Segment Routing Prefix
+ */
+static void sr_prefix_uninstall_label(const struct sr_prefix *srp)
+{
+	struct zapi_labels zl;
+
+	/* Prepare message. */
+	memset(&zl, 0, sizeof(zl));
+	zl.type = ZEBRA_LSP_ISIS_SR;
+	zl.local_label = srp->input_label;
+
+	/* Also remove routes in the RIB for remote prefixes only. */
+	if (srp->type == ISIS_SR_PREFIX_REMOTE) {
+		SET_FLAG(zl.message, ZAPI_LABELS_FTN);
+		zl.route.prefix = srp->prefix;
+		zl.route.type = ZEBRA_ROUTE_ISIS;
+		zl.route.instance = 0;
+	}
+
+	/* Send message to zebra. */
+	(void)zebra_send_mpls_labels(zclient, ZEBRA_MPLS_LABELS_DELETE, &zl);
+}
+
+/**
+ * Process local Prefix-SID and install it if possible. Input label is
+ * computed before installing it in LFIB.
+ *
+ * @param srp	Segment Routing Prefix
+ *
+ * @return	0 on success, -1 otherwise
+ */
+static int sr_prefix_install_local(struct sr_prefix *srp)
+{
+	mpls_label_t input_label;
 
 	/*
-	 * No need to install LSP to local Prefix-SID unless the
+	 * No need to install Label for local Prefix-SID unless the
 	 * no-PHP option is configured.
 	 */
 	if (!CHECK_FLAG(srp->sid.flags, ISIS_PREFIX_SID_NO_PHP)
 	    || CHECK_FLAG(srp->sid.flags, ISIS_PREFIX_SID_EXPLICIT_NULL))
 		return -1;
 
-	if (IS_DEBUG_ISIS(DEBUG_SR)) {
-		zlog_debug("ISIS-SR (%s) installing Prefix-SID %pFX %s %u (%s)",
-			   area->area_tag, &srp->prefix,
-			   CHECK_FLAG(srp->sid.flags, ISIS_PREFIX_SID_VALUE)
-				   ? "label"
-				   : "index",
-			   srp->sid.value, circuit_t2string(srn->level));
-		zlog_debug("  nexthop self");
-	}
+	sr_debug("  |- Installing Prefix-SID %pFX - %u nexthop self",
+		 &srp->prefix, srp->sid.value);
```

Comment:
In the new log message we're not displaying the IS-IS area, level and whether the Prefix-SID is an index or an absolute label. I'd prefer to retain that level of information in the logs.

Also, the old log messages follow the same pattern used by the rest of the isisd code, which is the name of the debug category (e.g. `ISIS-Rte`, `ISIS-Spf`, `ISIS-BFD`) followed by the IS-IS area name. If possible I'd prefer to stick to that pattern in order to be consistent.

Reply:
OK. Understand. I will change the 'SR(%s):' by 'ISIS-SR(%s):' tag. But, when I used '  |-' is to indicate the log message depends from the above i.e. we are in the same function or group of functions. If you look to the log globally, these made them more consistent and readable. However, I understand that this could be corrupted if you activate more than SR log. I will see how I could keep SR log coherency while taking into account other logs.

Reply:
Done